src/rt/util/utf.d

```diff
@@ -750,8 +756,14 @@ wptr toUTF16z(in char[] s)
     wchar[] r;
     size_t slen = s.length;
 
-    r.length = slen + 1;
-    r.length = 0;
+    if (!__ctfe)
+    {
+        // Reserve still does a lot if slen is zero.
+        // Return early for that case.
+        if (0 == slen)
+            return &""w[0];// Terminating null is auto-included at end of string literals.
```

Comment:
This isn’t right. Return `””w.ptr`

Reply:
If I do that I get a complaint about the code not being `@safe`. I'll assume it's okay to mark `toUTF16z` as `@trusted` since `toUTF16` already is.

Reply:
Well, it's not `@safe`, as the intention is to return a null-terminated array, which is inherently unsafe. Not only that, but index 0 doesn't exist, so you potentially would have a `RangeError` thrown (it should be what happens, since that is the point of making you index the first element in `@safe` code).

Where is the code that is calling this, complaining that it's not `@safe`?

Reply:
Oh, I see, it was marked `@safe`. It shouldn't have been. ugh...

Reply:
I think for now, even though it's not really `@safe`, the best option may be to use:

```D
return &"\0"w[0];
```

Reply:
&""w[0] will trigger out of bounds in safe code IIRC.

Reply:
Well, it's likely this code is *never* triggered. Who is going to toUTF16z an empty string? ;)

Reply:
We could have done `return &"\0"w[0]`

Reply:
> Well, it's not @safe, as the intention is to return a null-terminated array, which is inherently unsafe.

I don't think there is a problem: a pointer to something known to exist (in the DATA segment or in GC memory) is fine in @safe code, as this compiles:
```void main() @safe
{
	static const char a = 'a';
	const char* p = &a;
}
```
You are not supposed to do any arithmetics on p and accessing *p should be fine. So "a".ptr could be accepted in @safe code, even "".ptr as literals are guaranteed to be zero-terminated. Am I missing something?

Reply:
It's technically `@safe`, but disingenuous. The point is to return a null-terminated string, not a pointer to a single `char`.

Reply:
> So "a".ptr could be accepted in @safe code

Just noticed that this was still accepted as deprecated in dmd 2.078, but no longer in master. I cannot find it  in https://dlang.org/deprecate.html, though.

Reply:
https://dlang.org/changelog/pending.html#ptr-safe-end-of-deprecation 
When I made https://github.com/dlang/dmd/pull/7688, I didn't realize it wasn't on the deprecations page. We really need to have more structure in our deprecation process :/

Reply:
> It's technically @safe, but disingenuous.

I agree. The syntax `&a[0]` is not nice, especially with literals, but it makes range checks kind of explicit. Let's hope the compiler elides them if possible.